<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import plotly.express as px

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
SELECT
    COUNT(id)
FROM vacancies
            '''

In [5]:
answer_3_1 = pd.read_sql_query(query_3_1, connection)
answer_3_1

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\2262139106.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_1 = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
SELECT
    COUNT(DISTINCT id)
FROM employers
            '''

In [7]:
answer_3_2 = pd.read_sql_query(query_3_2, connection)
answer_3_2

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1178497560.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
SELECT
    COUNT(DISTINCT id)
FROM areas
            '''

In [9]:
answer_3_3 = pd.read_sql_query(query_3_3, connection)
answer_3_3

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\4098136936.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
SELECT
    COUNT(DISTINCT id)
FROM industries
            '''

In [11]:
answer_3_4 = pd.read_sql_query(query_3_4, connection)
answer_3_4

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\3272017797.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

### Выводы по предварительному анализу данных

На основе проведенного анализа данных можно сделать следующие выводы:

1. **Количество вакансий в базе данных**:
   - Всего в БД содержится **49.197 вакансий**.
   - Это указывает на значительный объем данных, что позволяет проводить глубокий анализ рынка труда.

2. **Количество работодателей**:
   - В БД содержится **23501 работодателей**.
   - Такое количество работодателей свидетельствует о разнообразии компаний, представленных в базе.

3. **Количество регионов**:
   - Вакансии размещены в **1.362 регионах**.
   - Это говорит о широком географическом охвате, включающей как крупные города, так и менее населенные регионы.

4. **Количество сфер деятельности**:
   - В базе данных представлено **294 сферы деятельности**.

#### Общий вывод:
БД охватывает широкий спектр регионов и сфер деятельности, что ценно для анализа рынка труда.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query_4_1 = f'''
SELECT
    COUNT(v.id) AS cnt,
    a.name AS area
FROM vacancies v
JOIN areas a ON v.area_id=a.id
GROUP BY 2
ORDER BY 1 DESC
            '''

In [13]:
answer_4_1 = pd.read_sql_query(query_4_1, connection)
answer_4_1

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1030361834.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  answer_4_1 = pd.read_sql_query(query_4_1, connection)


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

In [14]:
top_10_cities = '''
SELECT
    COUNT(v.id) AS cnt,
    a.name AS area
FROM vacancies v
JOIN areas a ON v.area_id=a.id
GROUP BY a.name
ORDER BY cnt DESC
LIMIT 10
'''

df = pd.read_sql_query(top_10_cities, connection)

fig = px.bar(df, x='area', y='cnt', color='area', title='Топ-10 регионов по количеству вакансий',
             labels={'area': 'Регион', 'cnt': 'Количество вакансий'})
fig.show()

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1755640502.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(top_10_cities, connection)


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
query_4_2 = f'''
SELECT
    COUNT(id)
FROM vacancies
WHERE salary_from IS NOT NULL
OR salary_to IS NOT NULL
            '''

In [16]:
answer_4_2 = pd.read_sql_query(query_4_2, connection)
answer_4_2

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\4247165315.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
query_4_3 = f'''
SELECT
    ROUND(AVG(salary_from)) rnd_avg_low,
    ROUND(AVG(salary_to)) rnd_avg_high
FROM vacancies
            '''

In [18]:
answer_4_3 = pd.read_sql_query(query_4_3, connection)
answer_4_3

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1371674522.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



rnd_avg_low  rnd_avg_high
0      71065.0      110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
query_4_4 = f'''
SELECT
    COUNT(id),
    schedule,
    employment
FROM vacancies
GROUP BY 2,3
ORDER BY 1 DESC
            '''

In [20]:
answer_4_4 = pd.read_sql_query(query_4_4, connection)
answer_4_4

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1828156850.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
query_4_5 = f'''
SELECT
    experience AS "Требуемый опыт работы"
FROM vacancies
GROUP BY experience
ORDER BY COUNT(id)
            '''

In [22]:
answer_4_5 = pd.read_sql_query(query_4_5, connection)
answer_4_5

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\280175601.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Требуемый опыт работы
0           Более 6 лет
1             Нет опыта
2         От 3 до 6 лет
3    От 1 года до 3 лет

***

### Выводы по детальному анализу данных

На основе проведенного анализа данных можно сделать следующие выводы:

---

#### 1. **Распределение вакансий по регионам**:
   - Наибольшее количество вакансий сосредоточено в крупных городах:
     - **Москва**: 5 333 вакансий
     - **Санкт-Петербург**: 2 851 вакансий
     - **Минск**: 2 112 вакансий
     - **Новосибирск**: 2 006 вакансий
     - **Алматы**: 1 892 вакансий
   - В то же время в небольших населенных пунктах, таких как **Тарко-Сале**, **Новоаннинский**, **Бирск**, **Сасово** и **Анна**, количество вакансий минимально (по 1 вакансии).
   - **Вывод**: Рынок труда сильно сконцентрирован в крупных городах, что указывает на неравномерное распределение вакансий по регионам.

---

#### 2. **Анализ зарплат**:
   - Количество вакансий с указанной зарплатой (либо нижняя, либо верхняя граница): **24 073**.
   - Средние значения зарплат:
     - **Нижняя граница**: 71 065 ₽
     - **Верхняя граница**: 110 537 ₽
   - **Вывод**: Разрыв между нижней и верхней границами зарплат указывает на различия в опыте, квалификации и региональных особенностях.

---

#### 3. **График работы и тип занятости**:
   - Наиболее популярные условия работы:
     - **Полный день + Полная занятость**: 35 367 вакансий
     - **Удаленная работа + Полная занятость**: 7 802 вакансий
     - **Гибкий график + Полная занятость**: 1 593 вакансий
   - Менее популярные варианты:
     - **Вахтовый метод + Проектная работа**: 2 вакансии
     - **Сменный график + Проектная работа**: 1 вакансия
   - **Вывод**: Большинство вакансий предполагают полную занятость с традиционным графиком работы, однако удаленная работа также набирает популярность.

---

#### 4. **Требуемый опыт работы**:
   - Распределение вакансий по опыту работы (в порядке возрастания количества вакансий):
     1. **Более 6 лет**
     2. **Нет опыта**
     3. **От 3 до 6 лет**
     4. **От 1 года до 3 лет**
   - **Вывод**: На рынке труда востребованы как опытные специалисты (более 6 лет), так и кандидаты без опыта.

---

#### Общий вывод:
Данные показывают, что рынок труда характеризуется:
- Высокой концентрацией вакансий в крупных городах.
- Широким диапазоном предлагаемых зарплат.
- Преобладанием вакансий с полной занятостью и традиционным графиком работы.
- Разнообразием требований к опыту работы, что делает рынок доступным как для начинающих специалистов, так и для опытных специалистов.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
query_5_1 = f'''
(SELECT
    e.name
FROM employers e
JOIN vacancies v ON e.id=v.employer_id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC
LIMIT 1)

UNION ALL

(SELECT
    e.name
FROM employers e
JOIN vacancies v ON e.id=v.employer_id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC
OFFSET 4
LIMIT 1)
            '''

In [24]:
answer_5_1 = pd.read_sql_query(query_5_1, connection)
answer_5_1

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\3668485945.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0         Яндекс
1  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
query_5_2 = f'''
SELECT
    a.name AS area,
    COUNT(DISTINCT v.id) AS cnt_vac,
    COUNT(distinct e.id) AS cnt_emp
FROM areas a
LEFT JOIN vacancies v ON a.id = v.area_id
LEFT JOIN employers e ON a.id = e.area
WHERE v.area_id IS NULL
GROUP BY a.id
ORDER BY  3 DESC
LIMIT 1
            '''

In [26]:
answer_5_2 = pd.read_sql_query(query_5_2, connection)
answer_5_2

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\729581154.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



area  cnt_vac  cnt_emp
0  Россия        0      410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
query_5_3 = f'''
SELECT
  e.name,
  COUNT(DISTINCT v.area_id) AS cnt
FROM employers e
JOIN vacancies v ON e.id=v.employer_id
GROUP BY 1
ORDER BY 2 DESC
            '''

In [28]:
answer_5_3 = pd.read_sql_query(query_5_3, connection)
answer_5_3

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\3554497784.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0                      Яндекс  181
1                  Ростелеком  152
2                  Спецремонт  116
3      Поляков Денис Иванович   88
4                   ООО ЕФИН    71
...                       ...  ...
14761                  UniSol    1
14762            UNISTORY LLC    1
14763                   UNIT6    1
14764     United Distribution    1
14765              UNITEL LLC    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
query_5_4 = f'''
SELECT
    COUNT(*) cnt
FROM employers e
LEFT JOIN employers_industries ei ON e.id=ei.employer_id
WHERE ei.industry_id IS NULL
            '''

In [30]:
answer_5_4 = pd.read_sql_query(query_5_4, connection)
answer_5_4

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\3753089257.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
query_5_5 = f'''
SELECT
    e.name
FROM employers e
JOIN employers_industries ei ON e.id=ei.employer_id
JOIN industries i ON ei.industry_id=i.id
GROUP BY 1
HAVING COUNT(i.id) = 4
ORDER BY 1
OFFSET 2
LIMIT 1
            '''

In [32]:
answer_5_5 = pd.read_sql_query(query_5_4, connection)
answer_5_5

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\800831559.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  8419

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
query_5_6 = f'''
SELECT
    COUNT(ei.employer_id)
FROM employers_industries ei
JOIN industries i ON ei.industry_id=i.id
WHERE i.name = 'Разработка программного обеспечения'
            '''

In [34]:
answer_5_6 = pd.read_sql_query(query_5_6, connection)
answer_5_6

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\2173459197.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [35]:
# Импортируем библиотеки для выгрзуки городов-миллионников
import requests
from bs4 import BeautifulSoup 


url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)

page = BeautifulSoup(response.text, 'html.parser')
tabl = page.find("table", class_="standard")

mil_cities = []
table = tabl.find("tbody")
rows = table.find_all("tr")

for i in rows:
  tags = list(i.find("a"))
  mil_cities.append(tags[0]) 
mil_cities = mil_cities[1:]   

query_5_7 = f'''
(SELECT 
  a.name,
  count(v.id) cnt
FROM employers e
JOIN vacancies v ON e.id = v.employer_id
JOIN areas a ON v.area_id = a.id
WHERE (e.name = 'Яндекс')
            '''
# строка для фильтрации городов миллионников
row = len(mil_cities)
if row > 0:
  string = f" AND (( a.name = '{mil_cities[0]}')" # если есть хоть один город миллионник
  if row > 1:
    for s in mil_cities[1:]:
      string += f" OR (a.name = '{s}')" # для каждого следующего города из списка

query_5_7 += (
string + f'''
)
GROUP BY a.name
ORDER BY 2 DESC)
'''
+
'''
UNION  ALL
SELECT
  'Total' total,
  COUNT(v.id)
FROM employers e
JOIN vacancies v on e.id = v.employer_id
JOIN areas a on v.area_id = a.id
WHERE (e.name = 'Яндекс')'''
+string
+
'''
)
GROUP BY total
ORDER BY  2
'''
)


In [36]:
answer_5_7 = pd.read_sql_query(query_5_7, connection)
answer_5_7

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\883904668.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4            Казань   25
5    Ростов-на-Дону   25
6             Пермь   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

### Выводы по анализу работадателей

На основе проведенного анализа данных можно сделать следующие выводы:

---

#### 1. **Топ-5 работодателей по количеству вакансий в разных регионах**:
   - **Яндекс**: 1933 вакансия
   - **Ростелеком**: 491 вакансии
   - **Тинькофф**: 444 вакансии
   - **СБЕР**: 428 вакансии
   - **Газпром нефть**: 331 вакансии
   - **Вывод**: **Яндекс** является лидером по количеству вакансий, что указывает на активность указанной компании на рынке труда и высокую потребность в специалистах. Остальные в топ-5 имеют примерно одинаковое количество вакансий на рынке.

---

#### 2. **Регионы без вакансий**:
   - Наибольшее количество работодателей без вакансий сосредоточено в:
     - **Россия**: 410 работодателей
     - **Казахстан**: 207 работодателей
     - **Московская область**: 75 работодателей
     - **Краснодарский край**: 19 работодателей
     - **Ростовская область**: 18 работодателей
   - **Вывод**: В ряде регионов, включая Казахстан, наблюдается низкая активность на рынке труда, что может быть связано с экономическими или инфраструктурными ограничениями. Россия попала в этот список, оценочно, по принципу того, что работадатель не указывает страну, а выбирает конкретный регион трудоустройства.

---

#### 3. **Сфера деятельности работодателей**:
   - **Количество работодателей без указанной сферы деятельности**: 8419 вакансий, что составляет **17,11%** от общего числа вакансий.
   - **Количество работодателей с указанной сферой деятельности "Разработка ПО"**: 3553 вакансий (**7,22%**).
   - **Вывод**: Большая часть работодателей не указывает сферу деятельности, что затрудняет анализ рынка труда. Однако сфера **Разработки ПО** выделяется как одна из наиболее популярных, что подчеркивает высокий спрос на IT-специалистов.

---

#### 4. **Распределение вакансий Яндекса в городах-миллионниках**:
   - Наибольшее количество вакансий Яндекса сосредоточено в:
     - **Москва**: 54 вакансии
     - **Санкт-Петербург**: 42 вакансии
     - **Екатеринбург**: 39 вакансий
     - **Нижний Новгород**: 36 вакансий
     - **Новосибирск**: 35 вакансий
   - **Вывод**: **Яндекс** активно развивается в крупных городах, что подтверждает его стратегию по охвату регионов с высокой концентрацией населения и квалифицированных специалистов.

---

#### Общий вывод:
Данные показывают, что рынок труда характеризуется:
- Лидерством крупных компаний, таких как **Яндекс** и **Ростелеком**, в количестве вакансий.
- Низкой активностью в ряде регионов, особенно в Казахстане.
- Высоким спросом на IT-специалистов, что подтверждается популярностью сферы **Разработки ПО**.
- Активное присутствие **Яндекса** в городах-миллионниках, что указывает на стратегию компании по развитию в крупных регионах.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
query_6_1 = f'''
SELECT 
    COUNT(id) cnt
FROM vacancies
WHERE name ILIKE '%data%' OR
      name ILIKE '%данн%'
            '''

In [38]:
answer_6_1 = pd.read_sql_query(query_6_1, connection)
answer_6_1

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\401457554.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
query_6_2 = f'''
SELECT 
    COUNT(id) cnt
FROM 
    vacancies
WHERE 
    (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
    OR name ILIKE '%ML%' AND name NOT ILIKE '%HTML%'
    )
    AND
    (
    name ILIKE '%junior%'
    OR experience ILIKE 'нет опыта'
    OR employment ILIKE 'стажировка'
    )
            '''

In [40]:
answer_6_2 = pd.read_sql_query(query_6_2, connection)
answer_6_2

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\4140246696.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
query_6_3 = f'''
SELECT 
    COUNT(id) cnt
FROM 
    vacancies
WHERE 
    (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
    OR name ILIKE '%ML%' AND name NOT ILIKE '%HTML%'
    )
    AND
    (
    key_skills ILIKE '%SQL%'
    OR key_skills ILIKE '%postgres%'
    )
            '''

In [42]:
answer_6_3 = pd.read_sql_query(query_6_3, connection)
answer_6_3

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1417771425.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
query_6_4 = f'''
SELECT 
    COUNT(id) cnt
FROM 
    vacancies
WHERE 
    (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
    OR name ILIKE '%ML%' AND name NOT ILIKE '%HTML%'
    )
    AND
    (
    key_skills ILIKE '%python%'
    )

            '''

In [44]:
answer_6_4 = pd.read_sql_query(query_6_4, connection)
answer_6_4

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\1713433382.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
query_6_5 = f'''
SELECT 
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2)
FROM 
    vacancies
WHERE 
    (
    (name ILIKE '%data scientist%')
    OR (name ILIKE '%data science%')
    OR (name ILIKE '%исследователь данных%')
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    OR (name LIKE '%ML%' AND (name NOT ILIKE '%HTML%'))
    AND (key_skills IS NOT NULL)
    )
            '''

In [46]:
answer_6_5 = pd.read_sql_query(query_6_5, connection)
answer_6_5

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\252717723.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
query_6_6 = f'''
SELECT
  experience,
  ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0)
FROM vacancies
WHERE
(
    (name ILIKE '%data scientist%')
    OR (name ILIKE '%data science%')
    OR (name ILIKE '%исследователь данных%')
    OR (name LIKE '%ML%' AND  (name NOT ILIKE '%html%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
)
GROUP BY 1
ORDER BY 1
            '''

In [48]:
answer_6_6 = pd.read_sql_query(query_6_6, connection)
answer_6_6

C:\Users\Din\AppData\Local\Temp\ipykernel_9980\4219046839.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

На основе проведенного предметного анализа данных можно сделать следующие выводы:

---

#### 1. **Общее количество вакансий**:
   - Всего проанализировано **1771 вакансия**, относящаяся к данным, что составляет **3.6%** от общего числа вакансий.
   - **Вывод**: Рынок вакансий в области данных достаточно обширен, что указывает на высокий спрос на специалистов в этой сфере.

---

#### 2. **Вакансии для начинающих дата-саентистов**:
   - Количество вакансий для начинающих дата-саентистов: **51**.
   - **Вывод**: Несмотря на то, что вакансий для начинающих специалистов меньше, чем для опытных, такие возможности существуют, что делает рынок доступным для новичков.

---

#### 3. **Ключевые навыки для дата-саентистов**:
   - **SQL или Postgres** указаны в **229 вакансиях**.
   - **Python** встречается в **357 вакансиях**.
   - В среднем работодатели указывают **6-7 навыков** в одной вакансии.
   - **Вывод**: 
     - **Python** является наиболее востребованным навыком в области Data Science.
     - **SQL/Postgres** также играют важную роль, что подчеркивает необходимость владения базами данных.
     - Работодатели ожидают от кандидатов владения широким спектром навыков, что делает профессию дата-саентиста многогранной.

---

#### 4. **Зарплатные ожидания в зависимости от опыта**:
   - **Нет опыта**: Средняя зарплата — **74 643 ₽**.
   - **От 1 года до 3 лет**: Средняя зарплата — **139 675 ₽**.
   - **От 3 до 6 лет**: Средняя зарплата — **243 115 ₽**.
   - **Более 6 лет**: Зарплата не указана.
   - **Вывод**: 
     - Зарплата напрямую зависит от опыта работы, что подтверждает ценность профессионального роста в данной сфере.
     - Наиболее значительный скачок в зарплате наблюдается при переходе от уровня **1-3 года** к уровню **3-6 лет**.
     - Оценочно, в вакансиях для дата-саентистов с уровнем работы **более 6 лет** не указаны конкретные цифры в связи с тем, что **ЗП с таким опытом работы договорная**, и обсуждается непосредственно на собеседовании.

---

#### Общий вывод:
Данные показывают, что рынок вакансий для дата-саентистов характеризуется:
- Высоким спросом на специалистов с опытом работы, особенно в диапазоне **3-6 лет**.
- Важностью ключевых навыков, таких как **Python** и **SQL/Postgres**, которые являются обязательными для большинства вакансий.
- Наличием возможностей для начинающих специалистов, хотя их количество ограничено.
- Четкой зависимостью уровня зарплаты от опыта работы, что мотивирует специалистов к профессиональному развитию.

# Общий вывод по проекту

# Итоги исследования и обобщение выводов

На основе проведенного анализа данных можно сделать следующие обобщенные выводы:

---

## 1. **Общая характеристика рынка труда**
   - В базе данных содержится **49 197 вакансий** от **23 501 работодателя**, охватывающих **1 362 региона** и **294 сферы деятельности**.
   - Это свидетельствует о широком географическом и профессиональном разнообразии рынка труда.
   - **Вывод**: База данных представляет собой ценный ресурс для анализа рынка труда, охватывающий как крупные города, так и менее населенные регионы.

---

## 2. **Концентрация вакансий в крупных городах**
   - Наибольшее количество вакансий сосредоточено в крупных городах:
     - **Москва**: 5 333 вакансии
     - **Санкт-Петербург**: 2 851 вакансия
     - **Минск**: 2 112 вакансий
     - **Новосибирск**: 2 006 вакансий
     - **Алматы**: 1 892 вакансии
   - В небольших населенных пунктах количество вакансий минимально (по 1 вакансии).
   - **Вывод**: Рынок труда сильно сконцентрирован в крупных городах, что указывает на неравномерное распределение вакансий по регионам.

---

## 3. **Спрос на IT-специалистов**
   - **Яндекс**, **Ростелеком**, **Тинькофф**, **СБЕР** и **Газпром нефть** являются лидерами по количеству вакансий.
   - Сфера **разработки ПО** выделяется как одна из наиболее популярных, с **3 553 работодателями**, указывающими её в качестве сферы деятельности.
   - **Вывод**: IT-сектор остается одним из самых востребованных, с высоким спросом на специалистов в области данных и разработки ПО.

---

## 4. **Требования к навыкам и опыт работы**
   - **Python** и **SQL/Postgres** являются ключевыми навыками для дата-саентистов, встречаясь в **357** и **229** вакансиях соответственно.
   - В среднем работодатели указывают **6-7 навыков** в одной вакансии.
   - Зарплата напрямую зависит от опыта работы:
     - **Нет опыта**: 74 643 ₽
     - **1-3 года**: 139 675 ₽
     - **3-6 лет**: 243 115 ₽
   - **Вывод**: Работодатели ожидают от кандидатов широкого спектра навыков, а уровень зарплаты значительно увеличивается с ростом опыта.

---

## 5. **Вакансии для начинающих специалистов**
   - Для начинающих дата-саентистов доступно **51 вакансия**, что составляет небольшую долю от общего числа.
   - **Вывод**: Несмотря на ограниченное количество вакансий для новичков, такие возможности существуют, что делает рынок доступным для начинающих специалистов.

---

# Дополнительные исследования и прогнозы

## 1. **Каких вакансий много?**
   - Наибольшее количество вакансий сосредоточено в IT-секторе, особенно в сферах **разработки ПО** и **Data Science**.
   - Также наблюдается высокий спрос на специалистов в крупных компаниях, таких как **Яндекс**, **Ростелеком**, **Тинькофф** и **СБЕР**.

## 2. **Рынок труда в различных городах**
   - Крупные города, такие как **Москва**, **Санкт-Петербург** и **Новосибирск**, являются центрами концентрации вакансий.
   - В небольших городах и регионах количество вакансий значительно ниже, что может быть связано с экономическими и инфраструктурными ограничениями.

## 3. **Интересные направления для дальнейших исследований**
   - **Анализ динамики зарплат**: Изучение изменения уровня зарплат в зависимости от региона, опыта и сферы деятельности.
   - **Исследование удаленной работы**: Анализ популярности удаленной работы в разных регионах и сферах деятельности.
   - **Прогнозирование спроса на специалистов**: На основе текущих данных можно построить модели прогнозирования спроса на специалистов в различных сферах.

---

# Общий итог

Рынок труда характеризуется:
- **Высокой концентрацией вакансий** в крупных городах и IT-секторе.
- **Широким диапазоном требований** к навыкам и опыту работы.
- **Неравномерным распределением вакансий** по регионам, с низкой активностью в небольших городах.
- **Наличием возможностей** для начинающих специалистов.

Для более глубокого понимания рынка труда рекомендуется продолжить исследования, уделив внимание динамике зарплат, спросу на специалистов в различных сферах и популярности удаленной работы. Это позволит сделать более точные прогнозы.